### PR TITLE
Message forwarding and load balancing

### PR DESCRIPTION
**Pull request by ross-pure**
This PR adds two features: message forwarding and load balancing.

### Motivation
These features address uses of `phi` that will come up commonly; passing a message on to a different task that will do the actual reducing, and improving performance by load balancing work across many workers.

#### Message Forwarding
Forwarding messages has been implemented through a more opinionated design using `Resolver`s. This corresponds to a special kind of `Sender` (note: not a `Task`) that given a mapping `Resolve(Message) Sender` will forward any messages it receives appropriately. This means that there is no functionality exposed that would enable a normal `Task` to forward messages; to forward messages you need to create a dedicated `Resolver` that does only forwarding.

#### Load Balancing
Load balancing is implemented by using the existing `Task` code: when constructing a `Task`, instead of simply specifying the channel capacity, you now need to pass in an `Options` object that as well as having a `Cap` field also has a `Scale` field. The latter field determines whether or not the task will be load balanced; any number less than 2 (which will be the case if this field it not explicitly initialised when creating an `Options`) then the task will function as normal. Otherwise, the scale indicates how many workers will handle incoming messages. Note that in this case the `Reducer`s should not have any state, since they will not be copied for each worker, but each worker will simply call `Reduce` on the same object - trying to modify state in `Reduce` would not be thread safe.

### CODE REVIEW

File: task/task.go [hunk 1]
```diff
@@ -153,3 +186,30 @@ func flatten(message Message) Message {
...
+type router struct {
+	resolverMu *sync.Mutex
+	resolver   Resolver
+}
+
+// NewResolver returns a new sender that represents a resolver. The given
+// resolver determines how the sender routes messages; any message `m` that is
+// sent to this sender will be sent to the sender determined by the resolver
+// through `Resolve(m)`.
+func NewResolver(resolver Resolver) Sender {
```
Comment (loongy) [inline task/task.go]: Should this be called `NewRouter`? I feel like it makes more sense.

File: task/exmaples/lb/main.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	// Construct the load balancer and user tasks.
+	lbOpts := phi.Options{Cap: n, Scale: n}
+	lb := LB{}
+	lbTask := phi.New(lb, lbOpts)
+	userOpts := phi.Options{Cap: n}
+	user, finished := NewUser(lbTask, n)
+	userTask := phi.New(&user, userOpts)
+
+	// Run the tasks
+	done := context.Background()
```
Comment (loongy) [inline task/exmaples/lb/main.go]: Usually, this is called `ctx` not `done`.

**Review by loongy — DISMISSED** [review]
Good stuff, only some minor comments about nomenclature.

Looking through the examples, it feels like we should keep a close eye on production use and see if a `SendAsync` method on `Sender` would help reduce copy/paste.

It also feels like the `Router` should be able to return multiple `Senders`. This could be done by having a special implementation of the `Sender` interface that is checked by the `router` after calling `Resolve` (similar to how `Messages` are handled). An extra goroutine would need to be spawned to wait on all `Senders` and compress all returned `Messages` into one `Messages` object.

I will create some basic issues outlining these, and we can tackle them in the future.

File: task/exmaples/lb/main.go (hunk 2 above)
Comment (ross-pure) [reply task/exmaples/lb/main.go]: Fixed in fa5cf78.

File: task/task.go (hunk 1 above)
Comment (ross-pure) [reply task/task.go]: Fixed in fa5cf78.

**Comment by loongy** [comment]
Issues #5 and #6 have been created to reflect two of the desirable features that come to light because of this PR.

**Review by loongy — APPROVED** [review]

**Pull request merged**